packages/x-tree-view/src/internals/TreeViewProvider/TreeViewChildrenItemProvider.tsx

```diff
@@ -0,0 +1,91 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import { useTreeViewContext } from './useTreeViewContext';
+import type { UseTreeViewJSXItemsSignature } from '../plugins/useTreeViewJSXItems';
+import type { UseTreeViewItemsSignature } from '../plugins/useTreeViewItems';
+import type { UseTreeViewIdSignature } from '../plugins/useTreeViewId';
+
+export const TreeViewChildrenItemContext =
+  React.createContext<TreeViewChildrenItemContextValue | null>(null);
+
+if (process.env.NODE_ENV !== 'production') {
+  TreeViewChildrenItemContext.displayName = 'TreeViewChildrenItemContext';
+}
+
+interface TreeViewChildrenItemProviderProps {
+  itemId?: string;
+  children: React.ReactNode;
+}
+
+export function TreeViewChildrenItemProvider(props: TreeViewChildrenItemProviderProps) {
+  const { children, itemId = null } = props;
+
+  const { instance, rootRef } =
+    useTreeViewContext<
+      [UseTreeViewJSXItemsSignature, UseTreeViewItemsSignature, UseTreeViewIdSignature]
+    >();
+  const childrenIdAttrToIdRef = React.useRef<Map<string, string>>(new Map());
+
+  React.useEffect(() => {
+    if (!rootRef.current) {
+      return;
+    }
+
+    let idAttr: string | null = null;
+    if (itemId == null) {
+      idAttr = rootRef.current.id;
+    } else {
+      // Undefined during 1st render
+      const itemMeta = instance.getItemMeta(itemId);
+      if (itemMeta !== undefined) {
+        idAttr = instance.getTreeItemIdAttribute(itemId, itemMeta.idAttribute);
+      }
+    }
+
+    if (idAttr == null) {
+      return;
+    }
+
+    const previousChildrenIds = instance.getItemOrderedChildrenIds(itemId ?? null) ?? [];
+    const childrenElements = rootRef.current.querySelectorAll(
+      `${itemId == null ? '' : `*[id="${idAttr}"] `}[role="treeitem"]:not(*[id="${idAttr}"] [role="treeitem"] [role="treeitem"])`,
```

Comment:
Nitpick: Does the truthy ternary case make sense here? 🤔 
Wouldn't simply returning an empty array be the same?
P.S. Running `document.querySelectorAll('')` seems to throw and exception. 🙈

Reply:
The ternay does not create an empty selector, it only removes `*[id="${idAttr}"]`.

Selector with the root: `[role="treeitem"]:not(*[id="${idAttr}"] [role="treeitem"] [role="treeitem"])`
Selector with nested item: `*[id="${idAttr}"] [role="treeitem"]:not(*[id="${idAttr}"] [role="treeitem"] [role="treeitem"])`

Reply:
I can do the ternary outside of the selector if this is more readable:

`itemId == null ? rootRef.querySelector() : rootRef.querySelector()`

I did not do it because if I move `[role="treeitem"]:not(*[id="${idAttr}"] [role="treeitem"] [role="treeitem"])` to a string, I loose the IDE selector analysis which makes it harder to read, so we would have to duplicate this string in both selector.
Both approaches are good to me :+1:

Reply:
Whoops, sorry. 🙈 
Well, the readability isn't the best, as we can clearly see. 😆 
But if it's the cleanest overall solution, I retract my comment. 👌